runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.hpp

```diff
@@ -0,0 +1,291 @@
+   void swap();
+   void dup(int32_t size, int32_t delta);
+
+   void nop();
+
+   void shift(TR::DataType type, ShiftOperator op);
+
+   void conversion(TR::DataType fromType, TR::DataType toType);
+
+   void comparison(TR::DataType type, ComparisonOperator op);
+
+   void goto_(int32_t label);
+   void conditionalBranch(TR::DataType type, int32_t label, ConditionalBranchOperator op);
+
+   void return_(TR::DataType type, bool oneBit=false);
+
+   void new_();
+   void newarray();
+   void anewarray();
+   void multianewarray(int32_t dimension);
+
+   void arraylength();
+
+   void instanceof();
+   void checkcast();
+
+   void get(bool isStatic);
+   void put(bool isStatic);
+
+   void iinc(int32_t index, int32_t incVal);
+
+   void monitor(bool kind); //true: enter, false: exit
+
+   void switch_(bool kind); //true: lookup, false: table
+
+   void athrow();
+
+   void invoke(TR::MethodSymbol::Kinds kind);
+
+   TR::SymbolReference* getSymbolReference(int32_t cpIndex, TR::MethodSymbol::Kinds kind);
+
+   TR_CallSite* getCallSite(TR::MethodSymbol::Kinds kind, int32_t bcIndex, int32_t cpIndex);
+
+   /*** Methods for creating different types of abstract values ***/
+   TR::AbsValue* createObject(TR_OpaqueClassBlock* opaqueClass, TR_YesNoMaybe isNonNull);
+
+   TR::AbsValue* createNullObject();
+
+   TR::AbsValue* createArrayObject(TR_OpaqueClassBlock* arrayClass, TR_YesNoMaybe isNonNull, int32_t lengthLow, int32_t lengthHigh, int32_t elementSize);
+   TR::AbsValue* createStringObject(TR::SymbolReference* symRef, TR_YesNoMaybe isNonNull);
+
+   TR::AbsValue* createIntConst(int32_t value);
+   TR::AbsValue* createLongConst(int64_t value);
+
+   TR::AbsValue* createIntRange(int32_t low, int32_t high);
+   TR::AbsValue* createLongRange(int64_t low, int64_t high);
+
+   TR::AbsValue* createTopInt();
+   TR::AbsValue* createTopLong();
+
+   TR::AbsValue* createTopDouble();
+   TR::AbsValue* createTopFloat();
+
+   TR::AbsValue* createTopObject();
+
+   bool isNullObject(TR::AbsValue* v);
+   bool isNonNullObject(TR::AbsValue* v);
+
+   bool isArrayObject(TR::AbsValue* v);
+   bool isObject(TR::AbsValue* v);
+
+   bool isIntConst(TR::AbsValue* v);
+   bool isIntRange(TR::AbsValue* v);
+   bool isInt(TR::AbsValue* v);
+
+   bool isLongConst(TR::AbsValue* v);
+   bool isLongRange(TR::AbsValue* v);
+   bool isLong(TR::AbsValue* v);
+
+   TR::Compilation* comp() { return _comp; }
+   TR::Region& region() { return _region; }
+   TR::ValuePropagation* vp() { return _vp; }
+   bool _insideLoop;
+   bool _lastTimeThrough;
+   int32_t _callerIndex;
+   TR::Block* _block;
+   TR::AbsStackMachineState* _state;
+   TR::AbsValue** _returnValue;
+   TR::Compilation* _comp;
+   TR::Region& _region;
+   TR::ValuePropagation* _vp;
+   TR::ResolvedMethodSymbol* _callerMethodSymbol;
+   TR_ResolvedMethod* _callerMethod;
+   TR_J9ByteCodeIterator& _bci;
+   TR::AbsVisitor* _visitor;
+   TR::InliningMethodSummary* _inliningMethodSummary;
+   };
+
+class RegionIterator
```

Comment:
I think we create another iterator for region is because we need to traverse the sub-nodes in a reverse postorder in order to make sure all predecessors are visited before visiting any successor. This is because we need to check in `TR::AbsInterpreter::transferBlockStatesFromPredeccesors()` if the states of any predecessor can be passed to the successor.